File: mockets/java.old/us/ihmc/mockets/TSNRangeHandler.java
/**
 * The TSNRangeHandler class takes care of 
 *
 * @author Mauro Tortonesi
 */
package us.ihmc.mockets;

import java.io.PrintStream;

import java.util.logging.Logger;

class TSNRangeHandler implements PacketOperationListener
{
    TSNRangeHandler (boolean control, boolean sequenced) 
    {
        // make sure sequenced is false if control is true
        assert !control || !sequenced;

        _logger = Logger.getLogger ("us.ihmc.mockets");

        _firstNode = null;
        _lastNode = _firstNode;
        
        _control = control;
        _sequenced = sequenced;

        _numOfRanges = 0;
    }

    private void insertBefore (Range r, Node nextNode)
    {
        assert nextNode != null : "nextNode must not be null.";        
        assert _numOfRanges >= 0;
        
        Node prevNode = nextNode._prev;
        Node n = new Node (prevNode, nextNode, r);
        nextNode._prev = n;
        if (prevNode == null) {
            _firstNode = n;
        } else {
            prevNode._next = n;
        }

        ++_numOfRanges;
    }
    
    private void insertBack (Range r) 
    {
        assert _numOfRanges >= 0;

        Node n = new Node (_lastNode, null, r);
        if (_lastNode != null) {
            _lastNode._next = n;
        } else {
            assert _firstNode == null;
            _firstNode = n;
        }
        _lastNode = n;
        
        ++_numOfRanges;
    }

    private void insertFront (Range r) 
    {
        assert _numOfRanges >= 0;
        
        Node n = new Node (null, _firstNode, r);
        if (_firstNode != null) {
            _firstNode._prev = n;
        } else {
            assert _lastNode == null;
            _lastNode = n;
        }
        _firstNode = n;

        ++_numOfRanges;
    }

    private void removeNode (Node n) 
    {
        assert n != null : "n must not be null.";
        assert _numOfRanges > 0;
        
        Node prevNode = n._prev;
        Node nextNode = n._next;

        if (prevNode == null) {
            _firstNode = nextNode;
        } else {
            prevNode._next = nextNode;
        }
        
        if (nextNode == null) {
            _lastNode = prevNode;
        } else {
            nextNode._prev = prevNode;
        }
        
        --_numOfRanges;
    }
    
    /* returns true if the packet was processed, false if it was simply 
     * discarded because already present */
    public synchronized boolean packetProcessed (long seqNum) 
    {
        if (_firstNode == null) {
            // the list is empty, we can append a new range at the back
            insertBack (new Range(seqNum));
            return true;
        } else if (SequentialArithmetic.add(seqNum, 1) == _firstNode._data._begin) {
            // this sequence number immediately preceeds the TSN range of the first node
            --_firstNode._data._begin;
            return true;
        } else if (seqNum == SequentialArithmetic.add(_lastNode._data._end, 1)) {
            // this sequence number immediately follows the TSN range of the last node
            ++_lastNode._data._end;
            return true;
        } else if (SequentialArithmetic.lessThan(seqNum, _firstNode._data._begin)) {
            // this sequence number preceeds the first range node, so we have to create 
            // a new range and insert it at the beginning of the list
            insertFront (new Range(seqNum));
            return true;
        } else if (SequentialArithmetic.greaterThan(seqNum, _lastNode._data._end)) {
            // this sequence number follows the last range node, so we have to create 
            // a new range and insert it at the end of the list
            insertBack (new Range(seqNum));
            return true;
        } else {
            // find the right spot for this sequence number
            Node currNode = _firstNode;
            Node nextNode = _firstNode._next;
            while (currNode != null &&
                   SequentialArithmetic.lessThan(seqNum, currNode._data._end)) {
                boolean added = false;
                if (seqNum == SequentialArithmetic.add(currNode._data._end, 1)) {
                    // the sequence number immediately follows that of the current range node
                    ++currNode._data._end;
                    added = true;
                } else if (nextNode != null) {
                    if (SequentialArithmetic.add(seqNum, 1) == nextNode._data._begin) {
                        // the sequence number immediately preceeds the next range node
                        --nextNode._data._begin;
                        added = true;
                    } else if (SequentialArithmetic.greaterThan (seqNum, currNode._data._end) &&
                               SequentialArithmetic.lessThan (seqNum, nextNode._data._begin)) {
                        // the sequence number belongs to a new range that is in between the current and next range nodes
                        insertBefore (new Range(seqNum), nextNode);
                        return true;
                    }
                }
                
                // exit from the loop when the new TSN has been inserted in the range list
                if (added) {
                    // check to see if we need to do a merge
                    if (nextNode != null &&
                        SequentialArithmetic.add(currNode._data._end, 1) == nextNode._data._begin) {
                        // merge current and next range nodes
                        currNode._data._end = nextNode._data._end;
                        // we can safely use removeNode here
                        removeNode (nextNode);
                    }
                    return true;
                }
                
                // update currNode and nextNode for next loop cicle
                currNode = nextNode;
                if (nextNode != null) {
                    nextNode = nextNode._next;
                }
            }
        }
        
        return false;
    }
    
    synchronized void fillACKInformation (ACKInformation ai) 
    {
        int singlesCount = 0;
        int rangesCount = 0;
        
        if (_numOfRanges > 0) {
            ACKInfoBlockRanges ranges = new ACKInfoBlockRanges (_control, _sequenced);
            ACKInfoBlockSingles singles = new ACKInfoBlockSingles (_control, _sequenced);
            
            for (Node n = _firstNode; n != null; n = n._next) {
                Range r = n._data;
                if (r._begin == r._end) {
                    //_logger.info ("Adding single " + r._begin);
                    singles.addSingle (r._begin);
                    ++singlesCount;
                } else {
                    //_logger.info ("Adding range (" + r._begin + "," + r._end + ")");
                    ACKInfoRange air = new ACKInfoRange (r._begin, r._end);
                    ranges.addRange (air);
                    ++rangesCount;
                }
            }

            if (singlesCount > 0) {                
                assert singles != null;
                ai.addACKInfoBlock (singles);
            }
            if (rangesCount > 0) {
                assert ranges != null;
                ai.addACKInfoBlock (ranges);
            }
        }
    }

    void _dump (PrintStream os)
    {
        os.println ("_numOfRanges: " + _numOfRanges);
        for (Node ptr = _firstNode; ptr != null; ptr = ptr._next)
            ptr._data._dump (os);
    }

    Range peek()
    {
        if (_firstNode == null)
            return null;

        return _firstNode._data;
    }

    void removeFirstRange()
    {
        if (_firstNode != null) {
            removeNode (_firstNode);
        }
    }


    class Range
    {
        Range (long sequenceNumber) 
        {
            if (sequenceNumber < 0)
                throw new IllegalArgumentException ("Ranges endpoints must be non negative.");
            
            _begin = sequenceNumber;
            _end = sequenceNumber;
        }

        void _dump (PrintStream os) 
        {
            os.println ("range (" + _begin + "," + _end + ")");
        }

        long begin()
        {
            return _begin;
        }
        
        long end()
        {
            return _end;
        }
        
        private long _begin;
        private long _end;
    }

    private class Node
    {
        Node (Range data) 
        {
            _prev = null;
            _next = null;
            _data = data;
        }

        Node (Node prev, Node next, Range data) 
        {
            _prev = prev;
            _next = next;
            _data = data;
        }
        
        Node _next;
        Node _prev;
        Range _data;
    }

    private Logger _logger;
    private Node _firstNode;
    private Node _lastNode;
    private int _numOfRanges;
    private boolean _control;
    private boolean _sequenced;
}
/*
 * vim: et ts=4 sw=4
 */

File: mockets/cpp/TSNRangeHandler.cpp
/*
 * TSNRangeHandler.cpp
 * 
 * This file is part of the IHMC Mockets Library/Component
 * Copyright (c) 2002-2014 IHMC.
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 * 
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS
 * 252.227-7014(a)(12) (February 2014).
 * 
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

#include "TSNRangeHandler.h"
#include "PacketMutators.h"

#include "SequentialArithmetic.h"

//#include "stdio.h"

using namespace NOMADSUtil;

/*!!*/ // Need to fix bug where the cumulative TSN should not be here because it will cause a problem when
       // this class is used for the cancelled chunk

TSNRangeHandler::TSNRangeHandler (void)
{
    _pFirstNode = _pLastNode = NULL;
}

TSNRangeHandler::~TSNRangeHandler (void)
{
    while (_pFirstNode != NULL) {
        Node *pTempNode = _pFirstNode;
        _pFirstNode = _pFirstNode->pNext;
        delete pTempNode;
    }
    _pFirstNode = _pLastNode = NULL;
}

int TSNRangeHandler::addTSN (uint32 ui32TSN)
{
    bool bAddedTSN = false;
    if (_pFirstNode == NULL) {
        // There are no other nodes
        _pFirstNode = _pLastNode = new Node();
        _pFirstNode->ui32Begin = _pFirstNode->ui32End = ui32TSN;
        bAddedTSN = true;
    }
    else if ((ui32TSN + 1) == _pFirstNode->ui32Begin) {
        // This sequence number immediately preceeds that of the first range node
        _pFirstNode->ui32Begin--;
        bAddedTSN = true;
    }
    else if (ui32TSN == (_pLastNode->ui32End + 1)) {
        // This sequence number immediately follows that of the last range node
        _pLastNode->ui32End++;
        bAddedTSN = true;
    }
    else if (NOMADSUtil::SequentialArithmetic::lessThan (ui32TSN, _pFirstNode->ui32Begin)) {
        // This sequence number belongs to a range that preceeds the first range node
        Node *pNewNode = new Node();
        pNewNode->ui32Begin = pNewNode->ui32End = ui32TSN;
        pNewNode->pNext = _pFirstNode;
        _pFirstNode->pPrev = pNewNode;
        _pFirstNode = pNewNode;
        bAddedTSN = true;
    }
    else if (NOMADSUtil::SequentialArithmetic::greaterThan (ui32TSN, _pLastNode->ui32End)) {
        // This sequence number belongs to a range that follows the last range node
        Node *pNewNode = new Node();
        pNewNode->ui32Begin = pNewNode->ui32End = ui32TSN;
        pNewNode->pPrev = _pLastNode;
        _pLastNode->pNext = pNewNode;
        _pLastNode = pNewNode;
        bAddedTSN = true;
    }
    else {
        // Find the right spot for this sequence number
        Node *pCurrNode = _pFirstNode;
        Node *pNextNode = _pFirstNode->pNext;
        while (pCurrNode != NULL) {
            bool bCheckForMerge = false;
            if (ui32TSN == (pCurrNode->ui32End + 1)) {
                // The sequence number immediately follows that of the current range node
                pCurrNode->ui32End++;
                bCheckForMerge = true;
                bAddedTSN = true;
            }
            else if (pNextNode != NULL) {
                if ((ui32TSN + 1) == pNextNode->ui32Begin) {
                    // The sequence number immediately preceeds the next range node
                    pNextNode->ui32Begin--;
                    bCheckForMerge = true;
                    bAddedTSN = true;
                }
                else if (NOMADSUtil::SequentialArithmetic::greaterThan (ui32TSN, pCurrNode->ui32End) &&
                         NOMADSUtil::SequentialArithmetic::lessThan (ui32TSN, pNextNode->ui32Begin)) {
                    // The sequence number belongs to a new range that is in between the current and next range nodes
                    Node *pNewNode = new Node();
                    pNewNode->ui32Begin = pNewNode->ui32End = ui32TSN;
                    pNewNode->pPrev = pCurrNode;
                    pNewNode->pNext = pNextNode;
                    pCurrNode->pNext = pNewNode;
                    pNextNode->pPrev = pNewNode;
                    bAddedTSN = true;
                    break;
                }
            }
            // Check to see if we need to do a merge
            if (bCheckForMerge && (pNextNode != NULL)) {
                if ((pCurrNode->ui32End + 1) == pNextNode->ui32Begin) {
                    // Need to merge current and next range nodes
                    pCurrNode->ui32End = pNextNode->ui32End;
                    pCurrNode->pNext = pNextNode->pNext;
                    if (_pLastNode == pNextNode) {
                        // pNextNode was the last node - update the pLastNode pointer
                        _pLastNode = pCurrNode;
                    }
                    else {
                        // pNextNode is going away, so make the back pointer of the node
                        // following pNextNode point to pCurrNode
                        // NOTE: Since pNextNode is not the last note (as determined by the
                        // previous if test), there must be a node following pNextNode
                        pNextNode->pNext->pPrev = pCurrNode;
                    }
                    delete pNextNode;
                    break;
                }
            }
            pCurrNode = pNextNode;
            if (pNextNode) {
                pNextNode = pNextNode->pNext;
            }
        }
    }

    if (bAddedTSN) {
        return 1;
    }
    else {
        return 0;
    }
}

int TSNRangeHandler::appendTSNInformation (TSNChunkMutator *pTCM)
{
    Node *pCurrNode = _pFirstNode;
    // Append all ranges first
    if (pTCM->startAddingRanges()) {
        return -1;
    }
    while (pCurrNode != NULL) {
        if (pCurrNode->ui32Begin == pCurrNode->ui32End) {
            // Skip for now
        }
        else {
            if (pTCM->addRange (pCurrNode->ui32Begin, pCurrNode->ui32End)) {
                return -2;
            }
        }
        pCurrNode = pCurrNode->pNext;
    }
    if (pTCM->doneAddingRanges()) {
        return -3;
    }

    // Now append the individual TSNs
    if (pTCM->startAddingTSNs()) {
        return -4;
    }
    pCurrNode = _pFirstNode;
    while (pCurrNode != NULL) {
        if (pCurrNode->ui32Begin == pCurrNode->ui32End) {
            if (pTCM->addTSN (pCurrNode->ui32Begin)) {
                return -5;
            }
        }
        pCurrNode = pCurrNode->pNext;
    }
    if (pTCM->doneAddingTSNs()) {
        return -6;
    }

    return 0;
}

int TSNRangeHandler::freeze (ObjectFreezer &objectFreezer)
{
    // Go through the whole list of nodes
    Node *pCurrNode = _pFirstNode;
    while (pCurrNode != NULL) {
        // Insert a control char to signal that another node follows
        objectFreezer << (unsigned char) 1;
        objectFreezer.putUInt32 (pCurrNode->ui32Begin);
        objectFreezer.putUInt32 (pCurrNode->ui32End);
        
/*        printf ("ui32Begin %lu\n", pCurrNode->ui32Begin);
        printf ("ui32End %lu\n", pCurrNode->ui32End);*/
        
        pCurrNode = pCurrNode->pNext;
    }
    // Insert a control char to signal that there are no more data
    objectFreezer << (unsigned char) 0;
    
    return 0;
}

int TSNRangeHandler::defrost (ObjectDefroster &objectDefroster)
{
    uint32 ui32Begin;
    uint32 ui32End;
    unsigned char moreData;
    objectDefroster >> moreData;
    // Insert all nodes.
    // Add every node as last node because nodes were ordered when freezed
    while (moreData) {
        objectDefroster >> ui32Begin;
        objectDefroster >> ui32End;
        if (_pLastNode == NULL) {
            _pFirstNode = _pLastNode = new Node();
            _pFirstNode->ui32Begin = _pLastNode->ui32Begin = ui32Begin;
            _pFirstNode->ui32End = _pLastNode->ui32End = ui32End;
        }
        else {
            Node *pNewNode = new Node();
            pNewNode->ui32Begin = ui32Begin;
            pNewNode->ui32End = ui32End;
            pNewNode->pPrev = _pLastNode;
            _pLastNode->pNext = pNewNode;
            _pLastNode = pNewNode;
        }
        objectDefroster >> moreData;
    }
    
    Node *pCurrNode = _pFirstNode;
    while (pCurrNode != NULL) {
/*        printf ("ui32Begin %lu\n", pCurrNode->ui32Begin);
        printf ("ui32End %lu\n", pCurrNode->ui32End);*/
        
        pCurrNode = pCurrNode->pNext;
    }
    
    return 0;
}

SAckTSNRangeHandler::SAckTSNRangeHandler (void)
{
    _ui32CumulativeTSN = 0;
}

int SAckTSNRangeHandler::addTSN (uint32 ui32TSN)
{
    if (NOMADSUtil::SequentialArithmetic::greaterThanOrEqual (_ui32CumulativeTSN, ui32TSN)) {
        return 0;
    }

    int rc = TSNRangeHandler::addTSN (ui32TSN);

    // Check to see if the last cumulative acknowledgement must be updated
    if ((_ui32CumulativeTSN + 1) == _pFirstNode->ui32Begin) {
        _ui32CumulativeTSN = _pFirstNode->ui32End;
        Node *pTempNode = _pFirstNode;
        _pFirstNode = _pFirstNode->pNext;
        delete pTempNode;
        if (_pFirstNode == NULL) {
            // The list is empty
            _pLastNode = NULL;
        }
    }

    return rc;
}

int SAckTSNRangeHandler::freeze (ObjectFreezer &objectFreezer)
{
    objectFreezer.putUInt32 (_ui32CumulativeTSN);
    
	/*  
	printf ("SAckTSNRangeHandler\n");
    printf ("_ui32CumulativeTSN %lu\n", _ui32CumulativeTSN);
	*/
    
    // call the superclass method freeze
    TSNRangeHandler::freeze (objectFreezer);
    return 0;
}

int SAckTSNRangeHandler::defrost (ObjectDefroster &objectDefroster)
{
    objectDefroster >> _ui32CumulativeTSN;
    
/*    printf ("SAckTSNRangeHandler\n");
    printf ("_ui32CumulativeTSN %lu\n", _ui32CumulativeTSN);*/
    
    // call the superclass method defrost
    if (0 != TSNRangeHandler::defrost (objectDefroster)) {
        return -2;
    }
    return 0;
}

int CancelledTSNRangeHandler::deleteTSNsUpTo (uint32 ui32TSN)
{
    if (_pFirstNode == NULL) {
        return 0;
    }

    // As an optimization, first check if the TSN >= the TSN in the last node
    if (NOMADSUtil::SequentialArithmetic::lessThanOrEqual (_pLastNode->ui32End, ui32TSN)) {
        // The whole list can be deleted
        while (_pFirstNode != NULL) {
            Node *pTempNode = _pFirstNode;
            _pFirstNode = _pFirstNode->pNext;
            delete pTempNode;
        }
        _pFirstNode = _pLastNode = NULL;
        return 0;
    }

    // Starting at the beginning, remove nodes whose end TSN is less than or equal to the specified TSN
    while ((_pFirstNode) && (NOMADSUtil::SequentialArithmetic::lessThanOrEqual (_pFirstNode->ui32End, ui32TSN))) {
        Node *pTempNode = _pFirstNode;
        _pFirstNode = _pFirstNode->pNext;
        delete pTempNode;
        if (_pFirstNode == NULL) {
            // The list is empty
            _pLastNode = NULL;
        }
    }

    // The last possibility is that the first node's range includes the specified TSN
    // If so, the start TSN needs to be updated
    if ((_pFirstNode) && (NOMADSUtil::SequentialArithmetic::lessThanOrEqual (_pFirstNode->ui32Begin, ui32TSN))) {
        _pFirstNode->ui32Begin = ui32TSN + 1;
    }

    return 0;
}

ReceivedTSNRangeHandler::ReceivedTSNRangeHandler (void)
{
    _bAdded0TSN = false;
    _ui32CumulativeTSN = 0;
}

int ReceivedTSNRangeHandler::addTSN (uint32 ui32TSN)
{
    if (ui32TSN == 0) {
        _bAdded0TSN = true;
    }
    if (NOMADSUtil::SequentialArithmetic::greaterThanOrEqual (_ui32CumulativeTSN, ui32TSN)) {
        // The cumulative TSN already encompasses the new TSN, so nothing needs to be done
        return 0;
    }

    int rc = TSNRangeHandler::addTSN (ui32TSN);

    // Check to see if the last cumulative acknowledgement must be updated
    if ((_ui32CumulativeTSN + 1) == _pFirstNode->ui32Begin) {
        _ui32CumulativeTSN = _pFirstNode->ui32End;
        Node *pTempNode = _pFirstNode;
        _pFirstNode = _pFirstNode->pNext;
        delete pTempNode;
        if (_pFirstNode == NULL) {
            // The list is empty
            _pLastNode = NULL;
        }
    }

    return rc;
}

bool ReceivedTSNRangeHandler::alreadyReceived (uint32 ui32TSN)
{
    if (ui32TSN == 0) {
        return _bAdded0TSN;
    }
    if (NOMADSUtil::SequentialArithmetic::greaterThanOrEqual (_ui32CumulativeTSN, ui32TSN)) {
        // The cumulative TSN is greater than the specified TSN, so it has already been received
        return true;
    }
    // Search the list
    Node *pCurrNode = _pFirstNode;
    while (pCurrNode != NULL) {
        if (NOMADSUtil::SequentialArithmetic::lessThanOrEqual (pCurrNode->ui32Begin, ui32TSN) &&
            NOMADSUtil::SequentialArithmetic::greaterThanOrEqual (pCurrNode->ui32End, ui32TSN)) {
            // The specified TSN falls within the range contained in this node in the list
            return true;
        }
        pCurrNode = pCurrNode->pNext;
    }
    // No entries in the list were found that contained the specified TSN
    return false;
}

int ReceivedTSNRangeHandler::freeze (NOMADSUtil::ObjectFreezer &objectFreezer)
{
    objectFreezer.putUInt32 (_ui32CumulativeTSN);
    objectFreezer.putBool (_bAdded0TSN);
    
    // Call the superclass method freeze
    if (0 != TSNRangeHandler::freeze (objectFreezer)) {
        return -1;
    }
    return 0;
}

int ReceivedTSNRangeHandler::defrost (NOMADSUtil::ObjectDefroster &objectDefroster)
{
    objectDefroster >> _ui32CumulativeTSN;
    objectDefroster >> _bAdded0TSN;

    // Call the superclass method defrost
    if (0 != TSNRangeHandler::defrost (objectDefroster)) {
        return -1;
    }
    return 0;
}
